pkg/service/http_service.go

```diff
@@ -29,42 +33,86 @@ type HTTPService struct {
 	Logger                   *log.Entry
 }
 
+func (s *HTTPService) ServeHTTPS() {
+}
+
+func (s *HTTPService) tlsListener(l net.Listener) net.Listener {
+	// Load certificates.
+	certificate, err := loadTLSCertificate(s.HTTPServiceConfiguration.ServerCertPath,
+		s.HTTPServiceConfiguration.ServerKeyPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	config := &tls.Config{
+		Certificates: []tls.Certificate{*certificate},
+		Rand:         rand.Reader,
+		MinVersion:   tls.VersionTLS12,
+	}
+
+	tlsl := tls.NewListener(l, config)
+	return tlsl
+}
+
 func (s *HTTPService) Serve(ctx context.Context, eval eval.IEvaluator) error {
 	s.GRPCService.eval = eval
-	grpcServer := grpc.NewServer()
-	gen.RegisterServiceServer(grpcServer, s.GRPCService)
-
+	// Mux Setup
 	mux := runtime.NewServeMux(
 		runtime.WithErrorHandler(s.HTTPErrorHandler),
 	)
-	err := gen.RegisterServiceHandlerFromEndpoint(
+	var tlsCreds credentials.TransportCredentials
+	var dialOpts []grpc.DialOption
+	var err error
+	if s.HTTPServiceConfiguration.ServerCertPath != "" && s.HTTPServiceConfiguration.ServerKeyPath != "" {
+		tlsCreds, err = loadTLSCredentials(s.HTTPServiceConfiguration.ServerCertPath,
+			s.HTTPServiceConfiguration.ServerKeyPath)
+		if err != nil {
+			return err
+		}
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(tlsCreds))
+	} else {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+	// GRPC Setup
+	grpcServer := grpc.NewServer()
+	gen.RegisterServiceServer(grpcServer, s.GRPCService)
+	err = gen.RegisterServiceHandlerFromEndpoint(
 		context.Background(),
 		mux,
 		fmt.Sprintf("localhost:%d", s.HTTPServiceConfiguration.Port),
-		[]grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())},
+		// TODO: Add TLS here when we have a certificate
+		dialOpts,
 	)
 	if err != nil {
 		log.Fatal(err)
 	}
-
-	server := http.Server{
-		Handler:           mux,
-		ReadHeaderTimeout: 60 * time.Second,
-	}
-
+	// Net listener
 	l, err := net.Listen("tcp", fmt.Sprintf(":%d", s.HTTPServiceConfiguration.Port))
 	if err != nil {
 		log.Fatal(err)
 	}
-	m := cmux.New(l)
-
-	httpL := m.Match(cmux.HTTP1Fast())
-	grpcL := m.Match(cmux.HTTP2())
 
-	go func() { handleServiceError(server.Serve(httpL)) }()
-	go func() { handleServiceError(grpcServer.Serve(grpcL)) }()
-	go func() { handleServiceError(m.Serve()) }()
+	server := &http.Server{
+		Handler:           mux,
+		ReadHeaderTimeout: 60 * time.Second,
+	}
+	tcpm := cmux.New(l)
+	// We first match on HTTP 1.1 methods.
+	httpl := tcpm.Match(cmux.HTTP1Fast())
+	// If not matched, we assume that its TLS.
+	tlsl := tcpm.Match(cmux.Any())
+	tlsl = s.tlsListener(tlsl)
```

Comment:
This will `log.fatal` if no `ServerCertPath` / `ServerKeyPath` are provided.

Reply:
Yeah I just ran into this, it's quite hard to fix